### PR TITLE
Agents

### PR DESCRIPTION
**Pull request by alalamav**
- branched off #19 
- defines a base class for modular conversational agents
- refactors the server to use agents instead of models
- formatting fixes on the client (no explicit modifications otherwise)

### CODE REVIEW

File: py-server/default_model.py [hunk 1]
```diff
@@ -23,37 +24,38 @@
...
+    def _build_message_history(self, state):
         """
         Converts langchain message history to ChatMessage format
         so that we can always refer back to the direct API version.
         """
+        if not state.message_history:
+            return None
         messages = []
-        for message in message_history:
+        for message in state.message_history:
```
Comment (adamconnors) [inline py-server/default_model.py:53]: I wonder if this is the right time to ditch the direct API version and use langchain throughout so we don't need to do this message conversion anymore?

File: py-server/history_tutor/history_tutor.py [hunk 2]
```diff
@@ -136,50 +136,47 @@ def __init__(self):
...
-        response = self.chat_model.invoke(messages)
-        return response.content, world_state
+        messages.extend(state.message_history)
+        messages.append(HumanMessage(state.message))
+        return messages
 
-    def build_world_state(self):
+    def _init_world_state(self):
+        # TODO(alalama): update world state to a dict
         if USE_FAKE_WORLD_STATE:
```
Comment (adamconnors) [inline py-server/history_tutor/history_tutor.py:155]: no need to keep the FAKE_WORLD_STATE that was just me hacking around. Can remove it for now & if we need something more structured as a dev tool we can add it later.

**Review by adamconnors — APPROVED** [review]
Looks great, thanks !

I wondered if we should make the DefaultAgent use langchain so we don't have to convert the message types & to make things a bit cleaner?

One other thing to think about (separate PR) would be what the unit-test looks like for Agents... Right now I have a few things buried in history_tutor_test.py but I can imagine some shared lightweight test framework to make building these things easier.

File: py-server/history_tutor/history_tutor.py (hunk 2 above)
Comment (alalamav) [reply py-server/history_tutor/history_tutor.py]: done in fa40e6f.

File: py-server/default_model.py (hunk 1 above)
Comment (alalamav) [reply py-server/default_model.py]: removed in fa40e6f

**Review by adamconnors — APPROVED** [review]
lgtm

**Pull request merged**